File: backend/app/models/history.py
from datetime import datetime, timezone

from sqlalchemy import DateTime, ForeignKey, Index, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.session import Base


class TicketHistory(Base):
    __tablename__ = "ticket_history"
    __table_args__ = (
        Index("ix_ticket_history_ticket_created_at", "ticket_id", "created_at"),
    )

    id: Mapped[int] = mapped_column(primary_key=True)
    ticket_id: Mapped[int] = mapped_column(ForeignKey("tickets.id"), nullable=False, index=True)
    actor_id: Mapped[int] = mapped_column(ForeignKey("users.id"), nullable=False, index=True)
    action: Mapped[str] = mapped_column(String(50), nullable=False, index=True)
    old_value: Mapped[str | None] = mapped_column(Text, nullable=True)
    new_value: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False, index=True
    )

    ticket = relationship("Ticket", back_populates="history")
    actor = relationship("User", back_populates="ticket_history")

File: frontend/lib/models/history.dart
import 'user.dart';

class TicketHistoryModel {
  final int id;
  final int ticketId;
  final int actorId;
  final String action;
  final String? oldValue;
  final String? newValue;
  final DateTime createdAt;
  final User? actor;

  const TicketHistoryModel({
    required this.id,
    required this.ticketId,
    required this.actorId,
    required this.action,
    required this.oldValue,
    required this.newValue,
    required this.createdAt,
    this.actor,
  });

  factory TicketHistoryModel.fromJson(Map<String, dynamic> json) {
    return TicketHistoryModel(
      id: json['id'],
      ticketId: json['ticket_id'],
      actorId: json['actor_id'],
      action: json['action'],
      oldValue: json['old_value'],
      newValue: json['new_value'],
      createdAt: DateTime.parse(json['created_at']),
      actor: json['actor'] != null ? User.fromJson(json['actor']) : null,
    );
  }
}
